### PR TITLE
Add new CloudStack module cs_physical_network

### PR DESCRIPTION
**Pull request by PatTheSilent**
##### SUMMARY
Add new CloudStack module cs_physical_network

Resubmitted #25647 with integration tests and adjustments to the new API

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
cs_physical_network

##### ADDITIONAL INFORMATION
The decision to extend `AnsibleCloudStack` with `self.physical_network` was based on the fact that `physicalnetworkid` just like `networkid` is a property not unique in the CloudStack API but common across multiple endpoints serving different purposes (like traffic types, VLAN IP ranges).

### CODE REVIEW

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (dpassante) [inline lib/ansible/modules/cloud/cloudstack/cs_physical_network.py]: I wonder if we shouldn't remove this check and let Cloudstack return the error itself or set the `tag` parameter type to string. Any thoughts?

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (dpassante) [inline lib/ansible/modules/cloud/cloudstack/cs_physical_network.py]: For consistency, please use the `delegate_to: localhost` syntax.

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (dpassante) [inline lib/ansible/modules/cloud/cloudstack/cs_physical_network.py]: `default: yes`

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (dpassante) [inline lib/ansible/modules/cloud/cloudstack/cs_physical_network.py]: Not required according to doc and arg spec.

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (dpassante) [inline lib/ansible/modules/cloud/cloudstack/cs_physical_network.py]: `required: false` is the default, can be omited.
Also, I would prefer to see the parameter types here.

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (dpassante) [inline lib/ansible/modules/cloud/cloudstack/cs_physical_network.py:247]: I think it's time to move this method to Cloudstack module utils.

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (dpassante) [inline lib/ansible/modules/cloud/cloudstack/cs_physical_network.py]: IMHO `default=None` is useless. Right?

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (dpassante) [inline lib/ansible/modules/cloud/cloudstack/cs_physical_network.py]: Wouldn't it be possible to add a choice list of possible Network Service Providers?

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (dpassante) [inline lib/ansible/modules/cloud/cloudstack/cs_physical_network.py]: Same here.

File: test/integration/targets/cs_physical_network/tasks/main.yml
Comment (dpassante) [inline test/integration/targets/cs_physical_network/tasks/main.yml:153]: Is there a way to see if enabled nsps are present?

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (dpassante) [inline lib/ansible/modules/cloud/cloudstack/cs_physical_network.py]: Please replace with the one-line license declaration.

**Review by dpassante — CHANGES_REQUESTED** [review]
@PatTheSilent Thanks for your contribution.
I haven't had the opportunity to try it. Also I have some questions about tags and nsps.

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (PatTheSilent) [reply lib/ansible/modules/cloud/cloudstack/cs_physical_network.py:247]: @dpassante I agree on it too. I wanted to do it in one of the previous PR's but it got quite messy. I want to put it in there because I've got pending changes in `cs_vlan_ip_range` and a new `cs_traffic_type` module that both need it. Although I would still need to override it in this module - the same case with `get_network` method and `cs_network` module.

You can check the code here: https://github.com/ansible/ansible/pull/54020/files#diff-96da40a9131629a9034804485b0d0808 I'll resubmit it once this PR is merged.

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (PatTheSilent) [reply lib/ansible/modules/cloud/cloudstack/cs_physical_network.py]: I'm quite new in the Cloudstack world so I really don't know how dynamic that list actually is between deployments. NSP validation is done by a separate call to the API. I'm not sure if it's a static list or a dynamic one depending on the installed/enabled plugins.

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (PatTheSilent) [reply lib/ansible/modules/cloud/cloudstack/cs_physical_network.py]: Setting the parameter type to string seems reasonable to me. I'll go ahead and implement it.

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (dpassante) [reply lib/ansible/modules/cloud/cloudstack/cs_physical_network.py]: OK, not a blocker anyway.

**Comment by resmo** [comment]
Would be nice if this would be in 2.8, "freeze" is next week.

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (PatTheSilent) [reply lib/ansible/modules/cloud/cloudstack/cs_physical_network.py]: Done.

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (PatTheSilent) [reply lib/ansible/modules/cloud/cloudstack/cs_physical_network.py]: Done.

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (PatTheSilent) [reply lib/ansible/modules/cloud/cloudstack/cs_physical_network.py]: Done.

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (PatTheSilent) [reply lib/ansible/modules/cloud/cloudstack/cs_physical_network.py]: Done.

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (PatTheSilent) [reply lib/ansible/modules/cloud/cloudstack/cs_physical_network.py]: Done.

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (PatTheSilent) [reply lib/ansible/modules/cloud/cloudstack/cs_physical_network.py]: Done.

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (PatTheSilent) [reply lib/ansible/modules/cloud/cloudstack/cs_physical_network.py]: Done.

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (PatTheSilent) [reply lib/ansible/modules/cloud/cloudstack/cs_physical_network.py]: Done.

File: test/integration/targets/cs_physical_network/tasks/main.yml
Comment (PatTheSilent) [reply test/integration/targets/cs_physical_network/tasks/main.yml:153]: @dpassante currently not but I'm thinking about something like this:

```diff
diff --git a/lib/ansible/modules/cloud/cloudstack/cs_physical_network.py b/lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
index 6fe7b6f54c..9699f6fccb 100644
--- a/lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
+++ b/lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
@@ -195,6 +195,52 @@ domain:
   returned: success
   type: str
   sample: domain1
+nsps:
+  description: Mapping of enabled or disabled Network Service Providers
+  type: complex
+  returned: on enabling/disabling of Network Service Providers
+  contains:
+    enabled:
+      description: Network Service Providers that were enabled
+      returned: on Network Service Provider enabling
+      type: dict
+      sample:
+      {
+        "VirtualRouter": {
+          "canenableindividualservice": true,
+          "id": "ed593eb9-3ba3-4f77-a5c9-72f28c86a156",
+          "name": "VirtualRouter",
+          "physicalnetworkid": "e54a4ee0-b1ee-4622-8f91-09048cee8315",
+          "servicelist": [
+            "Vpn",
+            "Dhcp",
+            "Dns",
+            "Gateway",
+            "Firewall",
+            "Lb",
+            "SourceNat",
+            "StaticNat",
+            "PortForwarding",
+            "UserData"
+          ],
+          "state": "Enabled"
+        }
+      }
+    disabled:
+      description: Network Service Providers that were disabled
+      returned: on Network Service Provider disabling
+      type: dict
+      sample:
+      {
+        "InternalLbVm": {
+          "canenableindividualservice": true,
+          "id": "26604aa3-ee87-4826-8572-04b7a25aaa3a",
+          "name": "InternalLbVm",
+          "physicalnetworkid": "e54a4ee0-b1ee-4622-8f91-09048cee8315",
+          "servicelist": ["Lb"],
+          "state": "Enabled"
+        }
+      }
 '''
 
 from ansible.module_utils.basic import AnsibleModule
@@ -437,14 +483,15 @@ def main():
     state = module.params.get('state')
     nsps_disabled = module.params.get('nsps_disabled', [])
     nsps_enabled = module.params.get('nsps_enabled', [])
+    nsps_result = dict()
 
     if state in ['absent']:
         network = acs_network.absent_network()
     else:
         network = acs_network.present_network()
     if nsps_disabled is not None:
-        for name in nsps_disabled:
-            acs_network.update_nsp(name=name, state='Disabled')
+        for nsp_name in nsps_disabled:
+            nsps_result['disabled'][nsp_name] = acs_network.update_nsp(name=nsp_name, state='Disabled')
 
     if nsps_enabled is not None:
         for nsp_name in nsps_enabled:
@@ -453,10 +500,11 @@ def main():
             elif nsp_name == 'InternalLbVm':
                 acs_network.set_loadbalancer_element_state(enabled=True, nsp_name=nsp_name)
 
-            acs_network.update_nsp(name=nsp_name, state='Enabled')
+            nsps_result['enabled'][nsp_name] = acs_network.update_nsp(name=nsp_name, state='Enabled')
 
     result = acs_network.get_result(network)
-
+    if nsps_result:
+        result['nsps'] = nsps_result
     module.exit_json(**result)
 
 
diff --git a/test/integration/targets/cs_physical_network/tasks/main.yml b/test/integration/targets/cs_physical_network/tasks/main.yml
index 66b8c755ab..72864a8cbf 100644
--- a/test/integration/targets/cs_physical_network/tasks/main.yml
+++ b/test/integration/targets/cs_physical_network/tasks/main.yml
@@ -126,6 +126,7 @@
       - pn.zone == cszone.name
       - pn.vlan == '100-200,300-400'
       - pn.state == 'Enabled'
+      - pn.nsps.enabled.InternalLbVm.physicalnetworkid == pn.id
 
 - name: ensure a network is disabled
   cs_physical_network:

```

Thoughts?

**Comment by PatTheSilent** [comment]
@dpassante @resmo I'm struggling a bit to get the tests to pass. They seem to be failing because some other tests don't clean up after themselves. I'll raise some PR's / issues once I find the problem.

File: test/integration/targets/cs_physical_network/tasks/main.yml
Comment (dpassante) [reply test/integration/targets/cs_physical_network/tasks/main.yml:153]: It seems a bit too verbose, not knowing if all this information will be useful. However, it would be nice to have at least one list of nsps enabled / disabled.
I'm wondering if `nsps_enabled` and` nsps_disabled` only returning a list of nsps would not be sufficient.

**Comment by dpassante** [comment]
@PatTheSilent @resmo My bad. I'll fix the problem.

File: test/integration/targets/cs_physical_network/tasks/main.yml
Comment (PatTheSilent) [reply test/integration/targets/cs_physical_network/tasks/main.yml:153]: Yeah, I think I slightly overdid it.

This should work:

```diff
diff --git a/lib/ansible/modules/cloud/cloudstack/cs_physical_network.py b/lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
index 6fe7b6f54c..1d37322bbc 100644
--- a/lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
+++ b/lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
@@ -195,6 +195,24 @@ domain:
   returned: success
   type: str
   sample: domain1
+nsps_enabled:
+  description: list of Network Service Providers that were enabled
+  returned: on Network Service Provider enabling
+  type: list
+  sample:
+   - VirtualRouter
+nsps_disabled:
+  description: list of Network Service Providers that were disabled
+  returned: on Network Service Provider disabling
+  type: list
+  sample:
+   - InternalLbVm
 '''
 
 from ansible.module_utils.basic import AnsibleModule
@@ -452,11 +471,13 @@ def main():
                 acs_network.set_vrouter_element_state(enabled=True, nsp_name=nsp_name)
             elif nsp_name == 'InternalLbVm':
                 acs_network.set_loadbalancer_element_state(enabled=True, nsp_name=nsp_name)

             acs_network.update_nsp(name=nsp_name, state='Enabled')
 

     result = acs_network.get_result(network)
+    result['nsps_enabled'] = nsps_enabled
+    result['nsps_disabled'] = nsps_disabled

     module.exit_json(**result)
 
 
diff --git a/test/integration/targets/cs_physical_network/tasks/main.yml b/test/integration/targets/cs_physical_network/tasks/main.yml
index 66b8c755ab..5394963b31 100644
--- a/test/integration/targets/cs_physical_network/tasks/main.yml
+++ b/test/integration/targets/cs_physical_network/tasks/main.yml
@@ -126,6 +126,7 @@
       - pn.zone == cszone.name
       - pn.vlan == '100-200,300-400'
       - pn.state == 'Enabled'
+      - 'InternalLbVm' in pn.nsps_enabled
 
 - name: ensure a network is disabled
   cs_physical_network:
```

Comment (dpassante) [reply test/integration/targets/cs_physical_network/tasks/main.yml:153]: LGTM

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (dpassante) [reply lib/ansible/modules/cloud/cloudstack/cs_physical_network.py:247]: Ok, I see.

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (dpassante) [inline lib/ansible/modules/cloud/cloudstack/cs_physical_network.py]:
~~~
- name: Ensure a network is present
  cs_physical_network:
~~~

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (dpassante) [inline lib/ansible/modules/cloud/cloudstack/cs_physical_network.py]: Not needed.

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (dpassante) [inline lib/ansible/modules/cloud/cloudstack/cs_physical_network.py]: Not needed. `required: false` should be removed for all params.

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (PatTheSilent) [reply lib/ansible/modules/cloud/cloudstack/cs_physical_network.py]: Done.

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (PatTheSilent) [reply lib/ansible/modules/cloud/cloudstack/cs_physical_network.py]: Done.

File: test/integration/targets/cs_physical_network/tasks/main.yml
Comment (PatTheSilent) [reply test/integration/targets/cs_physical_network/tasks/main.yml:153]: Done.

**Comment by PatTheSilent** [comment]
@dpassante thanks for fixing the problem with integration tests. I've rebased my branch and corrected (I think) all your suggestions.

**Comment by PatTheSilent** [comment]
ready_for_review

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (resmo) [inline lib/ansible/modules/cloud/cloudstack/cs_physical_network.py]: there is a missing condition for check mode

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (resmo) [inline lib/ansible/modules/cloud/cloudstack/cs_physical_network.py]: there is a missing condition for check mode

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (resmo) [inline lib/ansible/modules/cloud/cloudstack/cs_physical_network.py]: there is a missing condition for check mode

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (resmo) [inline lib/ansible/modules/cloud/cloudstack/cs_physical_network.py]: do you know if the name are case sensitive? AFAIK their are not, so we could change it in a way, any thoughts?

~~~python
if nsp_name.lower() in ['virtualrouter', 'vpcvirtualrouter']:
~~~

**Review by resmo — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (PatTheSilent) [reply lib/ansible/modules/cloud/cloudstack/cs_physical_network.py]: The names aren't used directly in API calls so it makes sense to make them case insensitive. Implemented in the latest commit.

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (PatTheSilent) [reply lib/ansible/modules/cloud/cloudstack/cs_physical_network.py]: Added. I've also added an integration test for it.

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (PatTheSilent) [reply lib/ansible/modules/cloud/cloudstack/cs_physical_network.py]: Added.

File: lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
Comment (PatTheSilent) [reply lib/ansible/modules/cloud/cloudstack/cs_physical_network.py]: Added.

**Review by resmo — APPROVED** [review]

**Comment by resmo** [comment]
@dpassante anything that blocks this from merging from your side?

shipit

**Review by dpassante — APPROVED** [review]
Good job @PatTheSilent !

shipit

**Comment by PatTheSilent** [comment]
bot_broken

**Comment by resmo** [comment]
(new modules won't get merged by the bot)
build failure unrelated. Merging. 
Thanks @PatTheSilent, well done!

**Pull request merged**
